libsnark/polynomial_commitments/kzg10_batched.tcc

```diff
@@ -0,0 +1,285 @@
+
+    // For convenience of variable naming, let $f_i \in fs$ and $g_i in gs$ be
+    // the two sets of polynomials. These are denoted $f_i$ and $f^\prime_i$ in
+    // [GWC19]. Similarly, $h_1$ and $h_2$ are used in place of $h$ and
+    // $h^\prime$ in the paper, and $\gamma_1$ and $\gamma_2$ in place of
+    // $\gamma$ and $\gamma^\prime$.
+    //
+    // Compute:
+    //
+    //   h_1(X) = \sum_{i=1}^{t_1} \gamma_1^{i-1} (f_i(X) - f_i(z_1)/(X - z_1)
+    //
+    //   h_2(X) = \sum_{i=1}^{t_2} \gamma_2^{i-1} (g_i(X) - g_i(z_1)/(X - z_1)
+    //
+    // then evaluate:
+    //
+    //   W_1 = [h_1(x)]_1
+    //   W_2 = [h_2(x)]_2
+
+    // Compute the sum of $\gamma_1^{i=1} f_i$ polynomials, and pass this
+    // polynomial into the kzg10 create_witness function, to yield $W_1$ and
+    // $sf_1$.
+
+    // TODO: Use the evaluations object to compute f_accum_eval and
+    // g_accum_eval, instead of evaluating the accumulated polynomial.
+
+    // Accumulate polynomials and get the witnesses
+    const polynomial<Field> f_accum =
+        internal::polynomial_accumulate_with_power_factors(fs, gamma_1);
+    const libff::Fr<ppT> f_accum_eval =
+        kzg10<ppT>::evaluate_polynomial(f_accum, z_1);
+    const libff::G1<ppT> f_accum_witness =
+        kzg10<ppT>::create_evaluation_witness(f_accum, z_1, f_accum_eval, srs);
+
+    const polynomial<Field> g_accum =
+        internal::polynomial_accumulate_with_power_factors(gs, gamma_2);
+    const libff::Fr<ppT> g_accum_eval =
+        kzg10<ppT>::evaluate_polynomial(g_accum, z_2);
+    const libff::G1<ppT> g_accum_witness =
+        kzg10<ppT>::create_evaluation_witness(g_accum, z_2, g_accum_eval, srs);
+
+    return evaluation_witness(f_accum_witness, g_accum_witness);
+}
+
+template<typename ppT>
+bool kzg10_batched_2_point<ppT>::verify_evaluations(
+    const Field &z_1,
+    const Field &z_2,
+    const evaluations &evaluations,
+    const srs &srs,
+    const Field &gamma_1,
+    const Field &gamma_2,
+    const evaluation_witness &witness,
+    const std::vector<commitment> &cm_1s,
+    const std::vector<commitment> &cm_2s,
+    const Field &r)
+{
+    // See Section 3, p13 of [GWC19].
+
+    const size_t t1 = cm_1s.size();
+    const size_t t2 = cm_2s.size();
+    assert(t1 == evaluations.s_1s.size());
+    assert(t2 == evaluations.s_2s.size());
+
+    // Compute:
+    //
+    //   F = \sum_{i=1}^{t1} \gamma_1^{i-1} (cm_1[i] - [s_1[i]]_1) +      (G)
+    //       r \sum_{i=1}^{t2} \gamma_2^{i-1} (cm_2[i] - [s_2[i]]_1)      (H)
+
+    const std::vector<Field> &s_1s = evaluations.s_1s;
+    const std::vector<Field> &s_2s = evaluations.s_2s;
+
+    // Compute:
+    //
+    //   s_1_accum  = \sum_i \gamma_1^{i-1} s_1[i]    (in the scalar field)
+    //   cm_1_accum = \sum_i \gamma_1^{i-1} cm_1[i]   (in G1)
+    //   G          = cm_1_accum - s_1_accum * G1::one()
+
+    Field s_1_accum = s_1s[t1 - 1];
+    libff::G1<ppT> cm_1_accum = cm_1s[t1 - 1];
+    for (size_t i = t1 - 2; i < t1; --i) {
```

Comment:
Are you relying on the overflow to stop the loop here?
Like `i` starts at `t1-2` and is decremented along the loop so the condition: `i < t1` always holds - trivially. But when the final iteration occurs: `i = 0` and thus when we re-enter and since `i` is a `size_t` we overflow and abort because now `i > t1`.
Why not just do `for (size_t i = t1 - 2; i >= 0; --i) {`?

Reply:
> I can't think of a better way when using size_t.

What about using a C-type cast here?

```cpp
for (size_t i = t1 - 2; i != (size_t)-1; --i) {
```

I find this more clear as it makes the intent more explicit.

Else, jdhao proposes a 2nd way in: https://jdhao.github.io/2017/10/07/loop-forward-backward-with-cpp-vector/

Reply:
For the other approaches, I'm not so keen on the `i != (size_t)-1` condition which uses equality instead of a less-than.  If the logic in the loop touches `i` (e.g. because it needs to skip some values, say), this can fail to terminate and crash.  The other solution from that link is cute, but takes a little bit more thought to determine exactly which values in the range are hit.

The other thing to consider is that compilers should warn about conditions such as `i >= 0` for `i` as `size_t` as a tortology.

I've added a comment to call out the use of the underflow.

Reply:
Ok no worries, the code as is is perfectly fine. That's a good point about the != vs inequality.

> I've added a comment to call out the use of the underflow.

Perfect.